drivers: platform: maxim: pwm: Add Maxim PWM platform driver

RaduSabau1 opened the pull request:
## Pull Request Description

This PR contains the commit adding the Maxim PWM driver.

 - Extra field from no_os_pwm_desc is used as the timer's prescaler used to calculate the ticks for both period and duty cycle.
 - The prescaler for both should be the same and is calculated depending on the duty cycle since it is always smaller than
    the period, therefor a good resolution of the PWM will be obtained.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

RaduSabau1 commented:
V2:

- Use mxc_tmr_cfg_t pointer variable as descriptor extra field and pass it in duty cycle and period setter/getter functions in order to use the prescaler.
- Dynamically allocate the memory for the cfg at init and also add error labels in case of failure to avoid memory leaks (same procedure goes for the remove function).

amiclaus commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+
+	cfg = (mxc_tmr_cfg_t *)no_os_calloc(1, sizeof(*cfg));
+	if (!cfg) {
+		return -ENOMEM;
```
this return should be dropped to fix the cppcheck build. the return is handled already on next line with the conditional jump

RaduSabau1 replied:
I am guessing just assigning ret = -ENOMEM, so the correct error code is returned.

dbogdan commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+/***************************************************************************//**
+ *   @file   maxim_pwm.c
+ *   @brief  Implementation of MAX32655 PWM driver.
```
Since this driver will be common to all supported Maxim MCUs, delete the mention of MAX32655.

dbogdan commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+}
+
+/**
+* @brief MAX32655 specific PWM ops
```
MAX32655 -> maxim

dbogdan commented on drivers/platform/maxim/common/maxim_pwm.h:
```diff
@@ -0,0 +1,58 @@
+};
+
+/**
+ * @brief MAX32655 specific PWM ops
```
MAX32655 -> maxim

dbogdan commented on drivers/platform/maxim/common/maxim_pwm.h:
```diff
@@ -0,0 +1,58 @@
+/***************************************************************************//**
+ *   @file   maxim_pwm.h
+ *   @brief  Header file of MAX32655 PWM driver.
```
MAX32655 -> Maxim

amiclaus commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+		*prescaler = TMR_PRES_2048;
+	else
+		*prescaler = TMR_PRES_4096;
+
```
nit: extra line.

amiclaus commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+ * @param prescaler - enum value to be provided to the init function.
+ * @return 0 in case of success, -EINVAL otherwise.
+ */
+static int _get_prescaler(uint32_t div, mxc_tmr_pres_t *prescaler)
```
this looks like the same as in the timer implementation. maybe we can somehow use that directly (just a thought)

amiclaus commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+ * @param desc - PWM descriptor.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t max_pwm_enable(struct no_os_pwm_desc *desc)
```
`int` type. check the entire code.

CiprianRegus commented on drivers/platform/maxim/common/maxim_pwm.c:
```diff
@@ -0,0 +1,355 @@
+
+	max_pwm_disable(desc);
+
+	ret = MXC_TMR_Init(MXC_TMR_GET_TMR(desc->id), &tmr_cfg, 1);
```
This function will overwrite the VDDIO level.

RaduSabau1 commented:
V3

- Added max_pwm_extra struct and using to store pin VDDIO and timer configuration and use along PWM API.
- Changed get_prescaler static function in order to get the best possible resolution out of the timer therefore being able to have smaller duty_cycles.

RaduSabau1 commented on drivers/platform/maxim/max32655/maxim_pwm.c:
```diff
@@ -0,0 +1,322 @@
+	i = MAX_PWM_TMR_RES;
+	cnt = 0;
+
+	while (div > NO_OS_BIT(i))
```
Indeed, "i" should be decremented as well.

RaduSabau1 commented on drivers/platform/maxim/max32655/maxim_pwm.c:
```diff
@@ -0,0 +1,322 @@
+				       NO_OS_DIV_ROUND_CLOSEST_ULL(NANO, duty_cycle_ns));
+
+	/* Set duty cycle to a minimum. */
+	if (!duty_ticks)
```
Thought that in case of the duty cycle that the user wants to set, a minimuk duty cycle should be set.

RaduSabau1 commented:
V3:

- Added maxim_pwm files for MAX32650, MAX32660 and MAX32665
- Modified for each MAX uController to support all available timers.
- Modified prescaler calculation function in order to obtain the maximum count possible without risking an overflow.

RaduSabau1 commented:
V4

- Fixed prescaler function.

RaduSabau1 commented:
V5

- Change the way real duty cycle value is computed.

RaduSabau1 commented:
V6

- When PWM is disabled it only stops the timer, and the timer comes to shut down at remove.
